File: lingting-security-grpc/src/test/java/live/lingting/framework/security/grpc/SecurityGrpcExpandConvert.java
package live.lingting.framework.security.grpc;

import live.lingting.framework.convert.SecurityGrpcConvert;
import live.lingting.framework.security.domain.AuthorizationVO;
import live.lingting.framework.security.domain.SecurityScope;
import live.lingting.framework.security.domain.SecurityScopeAttributes;
import live.lingting.framework.util.BooleanUtils;
import live.lingting.framework.util.CollectionUtils;
import org.mapstruct.Mapper;
import org.mapstruct.factory.Mappers;

/**
 * @author lingting 2024-01-30 20:19
 */
public class SecurityGrpcExpandConvert extends SecurityGrpcConvert {

	@Override
	public ExpandSecurityScope scopeExpand(SecurityScope scope) {
		ExpandSecurityScope expand = ExpandMapstruct.INSTANCE.of(scope);
		SecurityScopeAttributes attributes = scope.getAttributes();
		expand.setExpand(isExpand(attributes));
		return expand;
	}

	@Override
	public ExpandAuthorizationVO voExpand(AuthorizationVO vo) {
		ExpandAuthorizationVO expand = ExpandMapstruct.INSTANCE.of(vo);
		SecurityScopeAttributes attributes = vo.getAttributes();
		expand.setExpand(isExpand(attributes));
		return expand;
	}

	boolean isExpand(SecurityScopeAttributes attributes) {
		if (!CollectionUtils.isEmpty(attributes) && attributes.containsKey("expand")) {
			return BooleanUtils.isTrue(attributes.get("expand"));
		}
		return false;
	}

	@Mapper
	public interface ExpandMapstruct {

		ExpandMapstruct INSTANCE = Mappers.getMapper(ExpandMapstruct.class);

		ExpandAuthorizationVO of(AuthorizationVO vo);

		ExpandSecurityScope of(SecurityScope scope);

	}

}

File: lingting-security-grpc/src/test/kotlin/live/lingting/framework/security/grpc/SecurityGrpcExpandConvert.kt
package live.lingting.framework.security.grpc

import live.lingting.framework.security.domain.AuthorizationVO
import live.lingting.framework.security.domain.SecurityScope
import live.lingting.framework.security.domain.SecurityScopeAttributes
import live.lingting.framework.security.grpc.convert.SecurityGrpcConvert
import live.lingting.framework.util.BooleanUtils.isTrue

/**
 * @author lingting 2024-01-30 20:19
 */
class SecurityGrpcExpandConvert : SecurityGrpcConvert() {
    override fun scopeExpand(scope: SecurityScope): ExpandSecurityScope {
        val expand = ExpandSecurityScope()
        expand.from(scope)
        val attributes = scope.attributes
        expand.isExpand = isExpand(attributes)
        return expand
    }

    override fun voExpand(vo: AuthorizationVO): ExpandAuthorizationVO {
        val expand = ExpandAuthorizationVO()
        expand.from(vo)
        val attributes = vo.attributes
        expand.isExpand = isExpand(attributes)
        return expand
    }

    fun isExpand(attributes: SecurityScopeAttributes): Boolean {
        if (!attributes.isNullOrEmpty() && attributes.containsKey("expand")) {
            return attributes["expand"].isTrue()
        }
        return false
    }

}
